system/include/emscripten/threading.h

```diff
@@ -117,25 +121,94 @@ void *emscripten_sync_run_in_main_thread_7(int function, void *arg1, void *arg2,
 
 typedef void (*em_func_v)(void);
 typedef void (*em_func_vi)(int);
+typedef void (*em_func_vf)(float);
 typedef void (*em_func_vii)(int, int);
+typedef void (*em_func_vif)(int, float);
+typedef void (*em_func_vff)(float, float);
 typedef void (*em_func_viii)(int, int, int);
+typedef void (*em_func_viif)(int, int, float);
+typedef void (*em_func_viff)(int, float, float);
+typedef void (*em_func_vfff)(float, float, float);
+typedef void (*em_func_viiii)(int, int, int, int);
+typedef void (*em_func_vifff)(int, float, float, float);
+typedef void (*em_func_vffff)(float, float, float, float);
+typedef void (*em_func_viiiii)(int, int, int, int, int);
+typedef void (*em_func_viffff)(int, float, float, float, float);
+typedef void (*em_func_viiiiii)(int, int, int, int, int, int);
+typedef void (*em_func_viiiiiii)(int, int, int, int, int, int, int);
+typedef void (*em_func_viiiiiiii)(int, int, int, int, int, int, int, int);
+typedef void (*em_func_viiiiiiiii)(int, int, int, int, int, int, int, int, int);
 typedef int (*em_func_i)(void);
 typedef int (*em_func_ii)(int);
 typedef int (*em_func_iii)(int, int);
 typedef int (*em_func_iiii)(int, int, int);
 
-#define EM_FUNC_SIGNATURE int
-
-#define EM_FUNC_SIG_V 1024
-#define EM_FUNC_SIG_VI (1024 + 1)
-#define EM_FUNC_SIG_VII (1024 + 2)
-#define EM_FUNC_SIG_VIII (1024 + 3)
-#define EM_FUNC_SIG_I 2048
-#define EM_FUNC_SIG_II (2048 + 1)
-#define EM_FUNC_SIG_III (2048 + 2)
-#define EM_FUNC_SIG_IIII (2048 + 3)
-
-#define EM_FUNC_SIG_NUM_FUNC_ARGUMENTS(x) ((x) & 1023)
+// Encode function signatures into a single uint32_t integer.
+// N.B. This encoding scheme is internal to the implementation, and can change in the future. Do not depend on the
+// exact numbers in this scheme.
+#define EM_FUNC_SIGNATURE unsigned int
+
+// The encoding scheme is as follows:
+// - highest three bits identify the type of the return value
+#define EM_FUNC_SIG_RETURN_VALUE_MASK (0x7U << 29)
+
+#define EM_FUNC_SIG_RETURN_VALUE_V   0
+#define EM_FUNC_SIG_RETURN_VALUE_I   (0x1U << 29)
+#define EM_FUNC_SIG_RETURN_VALUE_I64 (0x2U << 29)
```

Comment:
No reason, just wanted to be explicit that this will always represent a 64-bit integer, and not subject to ABIs. C does not guarantee that `long long` is 64-bit?

Reply:
Yes, though pragmatically there have been far fewer recent platforms to have `int`s of different sizes than `long` or `long long`. The names 'i' and 'i64' are not something I invented for the purpose of this PR, but those are the tokens used across the toolchain, e.g. in parseTools, preamble and runtime to denote signatures of 32-bit and 64-bit integer sizes.

Reply:
It's not a big deal. If we remove the volatiles (and thus ensure that this code will be compatible with the upstream LLVM backend regardless of which behavior we end up with) then I'll be happy with this patch.